fedimint-api-client/src/api/mod.rs

```diff
@@ -912,123 +962,279 @@ impl WsFederationApi<WsClient> {
-            })
-            .unzip();
+impl ReconnectClientConnections {
+    pub fn new(connector: &DynClientConnector) -> Self {
+        ReconnectClientConnections {
+            connections: connector
+                .peers()
+                .into_iter()
+                .map(|peer| (peer, ClientConnection::new(peer, connector.clone())))
+                .collect(),
+        }
+    }
 
-        WsFederationApi {
-            peer_ids,
-            self_peer_id,
-            peers: Arc::new(peer_connections),
-            module_id: None,
+    async fn request(
+        &self,
+        peer: PeerId,
+        method: ApiMethod,
+        request: ApiRequestErased,
+    ) -> anyhow::Result<Value> {
+        let connection = self
+            .connections
+            .get(&peer)
+            .expect("Could not find client connection for peer {peer}")
+            .connection()
+            .await
+            .context("Peer is disconnected and we failed to reconnect")?;
+
+        if let Ok(response) = connection.request(method.clone(), request.clone()).await {
+            return Ok(response);
         }
+
+        sleep(Duration::from_millis(1000)).await;
+
+        self.connections
+            .get(&peer)
+            .expect("Could not find client connection for peer {peer}")
+            .connection()
+            .await
+            .context("Peer is disconnected")?
+            .request(method, request)
+            .await
     }
 }
 
-impl<C> FederationPeer<C>
-where
-    C: JsonRpcClient + 'static,
-{
-    #[instrument(target = LOG_CLIENT_NET_API, level = "trace", fields(peer = %self.peer_id, %method), skip_all)]
-    pub async fn request(&self, method: &str, params: &[Value]) -> JsonRpcResult<Value> {
-        // Strategies using timeouts often depend on failing requests returning quickly,
-        // so every request gets only one reconnection attempt.
-        const RETRIES: usize = 1;
-
-        for attempts in 0.. {
-            // The `match` statement below should always return if `RETRIES <= attempts`, so
-            // if we're looping again and `attempts` is greater than `RETRIES`, we have a
-            // bug.
-            debug_assert!(attempts <= RETRIES);
-
-            let rclient = self.client.read().await;
-            match rclient.client.get_try().await {
-                Ok(client) if client.is_connected() => {
-                    return client.request::<_, _>(method, params).await;
-                }
-                Err(err) => {
-                    if RETRIES <= attempts {
-                        return Err(JsonRpcClientError::Transport(err.into()));
+#[derive(Clone, Debug)]
+struct ClientConnection {
+    sender: async_channel::Sender<oneshot::Sender<DynClientConnection>>,
+}
+
+impl ClientConnection {
+    fn new(peer: PeerId, connector: DynClientConnector) -> ClientConnection {
+        let (sender, receiver) = bounded::<oneshot::Sender<DynClientConnection>>(128);
+
+        fedimint_core::task::spawn(
```

Comment:
For reference: this ignores panics. We could just sneak a `task_group` in here from the outside and `spawn` and not risk scratching head later. Though I don't know if that's a lot of complications or not...